File: sdk-ios/Tune/Tune/TunePowerHookManager.m
//
//  TunePowerHookManager.m
//  TuneMarketingConsoleSDK
//
//  Created by Matt Gowie on 7/27/15.
//  Copyright (c) 2015 Tune. All rights reserved.
//

#import "TunePowerHookManager.h"
#import "TuneManager.h"
#import "TuneSkyhookPayload.h"
#import "TuneStringUtils.h"
#import "TunePowerHookValue.h"
#import "TuneArrayUtils.h"
#import "TunePlaylist.h"
#import "TuneSkyhookCenter.h"
#import "TuneSkyhookPayloadConstants.h"
#import "TuneSkyhookConstants.h"
#import "TuneCallbackBlock.h"
#import "TunePowerHookExperimentDetails+Internal.h"
#import "TuneUtils.h"
#import "TuneState.h"

@implementation TunePowerHookManager

NSDictionary *_phookHash;

NSObject *phookDictLock;

NSArray *_phookChangedBlocks;

NSOperationQueue *powerhooksCallbackQueue;

#pragma mark - Initialization

- (id)initWithTuneManager:(TuneManager *)tuneManager {
    self = [super initWithTuneManager:tuneManager];
    
    if (self) {
        powerhooksCallbackQueue = [NSOperationQueue new];
        phookDictLock = [[NSObject alloc] init];
        [self reset];
    }
    
    return self;
}

- (void)bringUp {
    [self registerSkyhooks];
}

- (void)bringDown {
    [self unregisterSkyhooks];
    [self revertToDefaults:nil];
}

#pragma mark - Skyhook registration

- (void)registerSkyhooks {
    [self unregisterSkyhooks];
    
    [[TuneSkyhookCenter defaultCenter] addObserver:self selector:@selector(afterCurrentPlaylistChange:) name:TunePlaylistManagerCurrentPlaylistChanged object:nil priority:TuneSkyhookPriorityIrrelevant];
}

#pragma mark - Managment of Class

/* This is intentionally not exposed as part of the public headers, but it is used for tests and init */
- (void)reset {
    @synchronized(phookDictLock){
        _phookHash = [[NSDictionary alloc] init];
        _phookChangedBlocks = [[NSArray alloc] init];
    }
}

- (void)revertToDefaults:(TuneSkyhookPayload *)payload {
    // reset power hook variables to default values
    NSDictionary *values = [TunePowerHookManager getPowerHooksDictionary];
    for (NSString *hookId in values) {
        [self setValueForHookById:hookId value:[values[hookId] defaultValue]];
    }
}

#pragma mark - Access the PowerHooks

+ (TunePowerHookValue *)getPowerHook:(NSString *)key {
    TunePowerHookValue *result = nil;
    
    @synchronized(phookDictLock) {
        if (key != nil && _phookHash[key] != nil) {
            result = _phookHash[key];
        }
    }
    
    return result;
}

+ (NSDictionary *)getPowerHooksDictionary {
    NSDictionary *result = nil;
    
    @synchronized(phookDictLock) {
        result = _phookHash.copy;
    }
    
    return result;
}

- (NSArray *)getPowerHooks {
    @synchronized(phookDictLock) {
        return _phookHash.allValues.copy;
    }
}

+ (void)setPowerHook:(TunePowerHookValue *)value forKey:(NSString *)key {
    @synchronized(phookDictLock) {
        NSMutableDictionary *updatedPowerHookHash = _phookHash.mutableCopy;
        [updatedPowerHookHash setValue:value forKey:key];
        _phookHash = [NSDictionary dictionaryWithDictionary:updatedPowerHookHash];
    }
}

#pragma mark - Registering Power Hooks


- (void)registerHookWithId:(NSString *)hookId friendlyName:(NSString *)friendlyName defaultValue:(NSString *)defaultValue description:(NSString *)description approvedValues:(NSArray *)approvedValues {
    
    NSString *cleanHook = [TuneStringUtils scrubNameForMongo:hookId];
    
    if ([TuneStringUtils isBlank:cleanHook] || [TuneStringUtils isBlank:friendlyName] || defaultValue == nil) {
        ErrorLog(@"hookId and friendlyName must not be empty and default value cannot be nil. Not registering hookId:%@ friendlyName:%@ with defaultValue:%@", hookId, friendlyName, defaultValue);
        return;
    } else if (approvedValues != nil && approvedValues.count < 1) {
        ErrorLog(@"Attempted to register Tune Power Hook %@ with an empty approvedValues array. The approvedValues array must not be empty as this wouldn't allow any changes on the TMC side.", cleanHook);
        return;
    } else if (approvedValues != nil && ![TuneArrayUtils areAllElementsOfArray:approvedValues ofType:[NSString class]]) {
        ErrorLog(@"Attempted to register Tune Power Hook %@ with an approvedValues array that included a value with type other than NSString. The approvedValues array can only contain elements of type NSString.", cleanHook);
        return;
    }
    
    // If we've got a value during registration then that means we've loaded a power hook from disk.
    // We want to use that value and merge it with the values supplied during registration.
    NSString *currentValue = nil;
    if ([TunePowerHookManager getPowerHook:cleanHook]) {
        currentValue = [TunePowerHookManager getPowerHook:cleanHook].value;
    }
    
    
    NSMutableDictionary *dictionary = @{POWERHOOKVALUE_NAME:cleanHook,
                                        POWERHOOKVALUE_DEFAULT_VALUE:defaultValue,
                                        POWERHOOKVALUE_FRIENDLY_NAME:friendlyName}.mutableCopy;
    
    if (currentValue != nil) {
        [dictionary setObject:currentValue forKey:POWERHOOKVALUE_VALUE];
    } else {
        [dictionary setObject:defaultValue forKey:POWERHOOKVALUE_VALUE];
    }
    
    if (description != nil) {
        [dictionary setObject:description forKey:POWERHOOKVALUE_DESCRIPTION];
    }
        
    if (approvedValues != nil) {
        [dictionary setObject:approvedValues forKey:POWERHOOKVALUE_APPROVED_VALUES];
    }
        
    TunePowerHookValue *powerHookValue = [[TunePowerHookValue alloc] initWithDictionary:[NSDictionary dictionaryWithDictionary:dictionary]];
    [TunePowerHookManager setPowerHook:powerHookValue forKey:cleanHook];
}

- (NSString *)getValueForHookById:(NSString *)hookId {
    NSString *cleanHook = [TuneStringUtils scrubNameForMongo:hookId];
    NSString *value = @"";
    
    TunePowerHookValue *powerHookValue = [TunePowerHookManager getPowerHook:cleanHook];
    
    if(powerHookValue != nil) {
        value = powerHookValue.value;
    }
    
    return value;
}

#pragma mark - Setting Value For Power Hook

- (void)setValueForHookById:(NSString *)hookId value:(NSString *)value {
    NSString *cleanHook = [TuneStringUtils scrubNameForMongo:hookId];
    
    TunePowerHookValue *powerHookValue = [TunePowerHookManager getPowerHook:cleanHook];
    
    if (powerHookValue != nil) {
        TunePowerHookValue *newPowerHookValue = [powerHookValue cloneWithNewValue:value];
        
        [TunePowerHookManager setPowerHook:newPowerHookValue forKey:cleanHook];
    }
}

+ (BOOL)setValueForHookById:(NSString *)hookId withPlaylistDictionary:(NSDictionary *)dictionary {
    if ([TuneState isTMADisabled]) { return NO; }
    
    BOOL notifyOnChange = NO;
    
    TunePowerHookValue *existingPowerHookValue = [TunePowerHookManager getPowerHook:hookId];
    
    NSMutableDictionary *newPhookDict = dictionary.mutableCopy;
    if (existingPowerHookValue != nil) {
        @try {
            [newPhookDict setValue:hookId forKey:POWERHOOKVALUE_NAME];
            [newPhookDict setValue:existingPowerHookValue.defaultValue forKey:POWERHOOKVALUE_DEFAULT_VALUE];
            [newPhookDict setValue:existingPowerHookValue.friendlyName forKey:POWERHOOKVALUE_FRIENDLY_NAME];
            
            if (existingPowerHookValue.phookDescription != nil) {
                [newPhookDict setValue:existingPowerHookValue.phookDescription forKey:POWERHOOKVALUE_DESCRIPTION];
            }
            
            if (existingPowerHookValue.approvedValues != nil) {
                [newPhookDict setValue:existingPowerHookValue.approvedValues forKey:POWERHOOKVALUE_APPROVED_VALUES];
            }
        } @catch (NSException *exception) {
            ErrorLog(@"Failed to serialize new Power Hook Value. Exception: %@", [exception reason]);
        }
        
        // We create a new TunePowerHookValue and replace existing.
        // If we didn't do this and merged existing with new we could be in a weird state
        TunePowerHookValue *newPowerHookValue = [[TunePowerHookValue alloc] initWithDictionary:newPhookDict];
        [TunePowerHookManager setPowerHook:newPowerHookValue forKey:hookId];
        
        // If the values are not equal send a notification out that the variable has changed
        if(![newPowerHookValue.value isEqualToString:existingPowerHookValue.value]) {
            notifyOnChange = YES;
        }
    } else {
        
        // No Power Hook with this hookId has been registered yet. This is either due to the customer
        // removing the Power Hook from their App (fine to regsiter since they won't use) or we're loading
        // this playlist from disk and the AppDelegate call hasn't happened yet.
        TunePowerHookValue *newPowerHookValue = [[TunePowerHookValue alloc] initWithDictionary:newPhookDict];
        [TunePowerHookManager setPowerHook:newPowerHookValue forKey:hookId];
        
        // Being explicit here: We don't want a power hooks changed block to be fired.
        notifyOnChange = NO;
    }
    
    return notifyOnChange;
}

#pragma mark - Updating Power Hooks via Playlist

- (void)updatePowerHooksFromPlaylist:(TunePlaylist *)playlist playlistFromDisk:(BOOL)playlistFromDisk {
    // If TMA is disabled then don't update the powerhooks
    if ([TuneState isTMADisabled]) { return; }
    
    __block BOOL notifyOnPowerHookChanges = NO;
    
    NSDictionary *powerHookSingleValues = playlist.powerHooks;
    [powerHookSingleValues enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
        if([TunePowerHookManager setValueForHookById:key withPlaylistDictionary:obj]){
            notifyOnPowerHookChanges = YES;
        }
    }];
    
    if(!playlistFromDisk){
      if(notifyOnPowerHookChanges){
          [self executeOnPowerHooksChangedBlocks];
      }
    }
}

- (void)afterCurrentPlaylistChange:(TuneSkyhookPayload *)payload {
    TunePlaylist *activePlaylist = [payload userInfo][TunePayloadNewPlaylist];
    
    NSNumber *playlistFromDiskNumber = ((NSNumber *)[payload userInfo][TunePayloadPlaylistLoadedFromDisk]);
    BOOL playlistFromDisk = (playlistFromDiskNumber != nil) && ([playlistFromDiskNumber boolValue]);
    
    [self updatePowerHooksFromPlaylist:activePlaylist playlistFromDisk:playlistFromDisk];
}

#pragma mark - Power Hooks Changed

- (void)onPowerHooksChanged:(void (^)())block {
    TuneCallbackBlock *blockCallback = [[TuneCallbackBlock alloc] initWithCallbackBlock:block fireOnce:NO];
    
    @synchronized(phookDictLock){
        NSMutableArray *updatedPowerHooksChangedBlocks = _phookChangedBlocks.mutableCopy;
        [updatedPowerHooksChangedBlocks addObject:blockCallback];
        _phookChangedBlocks = [NSArray arrayWithArray:updatedPowerHooksChangedBlocks];
    }
}

- (void)executeOnPowerHooksChangedBlocks {
    NSArray *callbacks;
    
    @synchronized(phookDictLock) {
        callbacks = _phookChangedBlocks.copy;
    }
    
    for (TuneCallbackBlock *callback in callbacks) {
        [powerhooksCallbackQueue addOperationWithBlock:^{
            [callback executeBlock];
        }];
    }
}

#pragma mark - Testing Helpers

+ (NSDictionary *)getSingleValuePowerHooks {
    @synchronized(phookDictLock){
        NSMutableDictionary *allSingleValues = [[NSMutableDictionary alloc] init];
        
        NSDictionary *values = [TunePowerHookManager getPowerHooksDictionary];
        for (NSString *hookId in values) {
            [allSingleValues setValue:@{ POWERHOOKVALUE_FRIENDLY_NAME : [values[hookId] friendlyName],
                                         POWERHOOKVALUE_DEFAULT_VALUE : [values[hookId] defaultValue] } forKey:hookId];
        }
        
        return allSingleValues;
    }
}

@end

File: sdk-android/TuneMarketingConsoleSDK/src/main/java/com/tune/ma/powerhooks/TunePowerHookManager.java
package com.tune.ma.powerhooks;

import com.tune.TuneDebugLog;
import com.tune.ma.model.TuneCallback;
import com.tune.ma.playlist.model.TunePlaylist;
import com.tune.ma.powerhooks.model.TunePowerHookValue;
import com.tune.ma.utils.TuneJsonUtils;
import com.tune.ma.utils.TuneStringUtils;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Created by gowie on 1/25/16.
 */
public class TunePowerHookManager {

    private Map<String, TunePowerHookValue> phookHash;
    private TuneCallback callback;
    private ExecutorService executorService;

    /**
     * Set of Power Hook IDs explicitly registered by the user.
     */
    private Set<String> userRegisteredPowerHooks;

    public TunePowerHookManager() {
        phookHash = new HashMap<String, TunePowerHookValue>();
        callback = null;
        userRegisteredPowerHooks = new HashSet<String>();
        this.executorService = Executors.newSingleThreadExecutor();
    }

    // Updating From Playlist
    //////////////////////////

    public void updatePowerHooksFromPlaylist(TunePlaylist playlist) {
        // TODO: Check disabled

        boolean notifyPowerHooksChanged = false;

        JSONObject phookJson = playlist.getPowerHooks();
        if (phookJson == null) {
            return;
        }

        Iterator<String> phookIds = phookJson.keys();
        while (phookIds.hasNext()) {
            String hookId = phookIds.next();

            if (this.mergeInPlaylistPowerHook(hookId, TuneJsonUtils.getJSONObject(phookJson, hookId))) {
                notifyPowerHooksChanged = true;
            }
        }

        if (!playlist.isFromDisk() && notifyPowerHooksChanged) {
            this.executeOnPowerHooksChangedBlocks();
        }
    }

    private boolean mergeInPlaylistPowerHook(String hookId, JSONObject json) {
        boolean notifyChange = false;

        TunePowerHookValue existingPhook = getPowerHookValue(hookId);
        TunePowerHookValue newPhook = null;
        if (existingPhook != null) {

            try {
                newPhook = existingPhook.clone();
                newPhook.mergeWithPlaylistJson(json);

                phookHash.put(hookId, newPhook);

                // If the values are not equal send a notification out that the Phook has changed
                if (!newPhook.getValue().equals(existingPhook.getValue())) {
                    notifyChange = true;
                }
            } catch (CloneNotSupportedException e) {
                e.printStackTrace();
                TuneDebugLog.e("Failed to clone existingPhook: " + existingPhook.getHookId());
            }
        } else {

            // No Power Hook with this hookId has been registered yet. This is either due to the customer
            // removing the Power Hook from their App (fine to register since they won't use) or we're loading
            // this playlist from disk and the Application#registerPowerHooks call hasn't happened yet.
            newPhook = new TunePowerHookValue();
            newPhook.setHookId(hookId);
            newPhook.mergeWithPlaylistJson(json);
            phookHash.put(hookId, newPhook);

            notifyChange = false;
        }

        return notifyChange;
    }

    private void executeOnPowerHooksChangedBlocks() {
        executorService.execute(new Runnable() {
            @Override
            public void run() {
                if (callback != null) {
                    callback.execute();
                }
            }
        });
    }

    // Public API
    //////////////

    public synchronized void registerPowerHook(String hookId, String friendlyName, String defaultValue, String description, List<String> approvedValues) {
        if (hookId == null || friendlyName == null || defaultValue == null) {
            TuneDebugLog.IAMConfigError("TUNE Power Hook IDs, friendly names and default values cannot be null. This registration " +
                    "(hookId:" + hookId + ", friendlyName:" + friendlyName + ", defaultValue: " + defaultValue + ") will be ignored.");
            return;
        }

        String scrubbedHookId = TuneStringUtils.scrubStringForMongo(hookId);

        // do not allow an existing Power Hook ID to be overwritten
        if (userRegisteredPowerHooks.contains(scrubbedHookId)) {
            TuneDebugLog.IAMConfigError("Invalid attempt to overwrite a previously registered Power Hook for hook ID \"" + hookId + "\".");
        } else {
            userRegisteredPowerHooks.add(scrubbedHookId);

            TunePowerHookValue existingPowerHook = getPowerHookValue(hookId);
            if (existingPowerHook != null) {
                existingPowerHook.setFriendlyName(friendlyName);
                existingPowerHook.setDefaultValue(defaultValue);
                existingPowerHook.setDescription(description);
                existingPowerHook.setApprovedValues(approvedValues);
            } else {
                phookHash.put(scrubbedHookId, new TunePowerHookValue(scrubbedHookId, friendlyName, defaultValue, description, approvedValues));
            }
        }
    }

    public synchronized String getValueForHookById(String hookId) {
        String scrubbedHookId = TuneStringUtils.scrubStringForMongo(hookId);
        TunePowerHookValue phookValue = phookHash.get(scrubbedHookId);

        if (phookValue == null) {
            TuneDebugLog.IAMConfigError("No Power Hook was registered with the given Hook ID: " + hookId);
            return null;
        } else {
            return phookValue.getValue();
        }
    }

    public synchronized void setValueForHookById(String hookId, String value) {
        TunePowerHookValue phookValue = getPowerHookValue(hookId);

        if (phookValue == null) {
            TuneDebugLog.IAMConfigError("No Power Hook was registered with the given Hook ID: " + hookId);
        } else {
            phookValue.setValue(value);
        }
    }

    public synchronized void onPowerHooksChanged(TuneCallback callback) {
        this.callback = callback;
    }

    public synchronized List<TunePowerHookValue> getPowerHookValues() {
        List<TunePowerHookValue> values = new ArrayList<TunePowerHookValue>();
        for (Map.Entry<String, TunePowerHookValue> entry: phookHash.entrySet()) {
            values.add(entry.getValue());
        }
        return values;
    }

    // Helpers
    ///////////

    private TunePowerHookValue getPowerHookValue(String hookId) {
        String scrubbedHookId = TuneStringUtils.scrubStringForMongo(hookId);
        TunePowerHookValue phookValue = phookHash.get(scrubbedHookId);
        return phookValue;
    }

    protected Map<String, TunePowerHookValue> getPowerHooks() {
        return new HashMap<String, TunePowerHookValue>(phookHash);
    }

    protected void clearPowerHooks() {
        phookHash = new HashMap<String, TunePowerHookValue>();
    }

    // package private
    void setExecutorService(ExecutorService executorService) {
        this.executorService = executorService;
    }
}
